### PR TITLE
Now mapping Aquarius result qualifier to the rdb column 'lev_status_cd'

### PR DESCRIPTION
**Pull request by ee-usgs**
* Aquarius uses multi-values for qualifiers but NWISWeb is single valued, single character.  One qual is choosen based on the order in the LevelStatusCode enum.
* siteStatusCode is not used for the mapping but was thought to be the correct AQ column - references to it were removed
* Added a new 'rules' class the does the mapping of AQ result code to NWISWeb lev_status_cd w/ the help of the LevelStatusCode enum
* Updated tests
* Updated the pom file to resolve some version conflicts and a missing version number

Before making a pull request
----------------------------

- [x] Make sure all tests run
- [x] Update the changelog appropriately

Title
-----------
Brief description of changes. Reference the JIRA ticket if appropriate

Description
-----------
If no JIRA ticket is referenced, describe the changes made. Note anything that you want the reviewers to know while
reviewing your pull request

After making a pull request
---------------------------
- [x] If appropriate, put the link to the PR in the JIRA ticket
- [x] Assign someone to review unless the change is trivial

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -68,6 +68,12 @@
             <artifactId>aws-lambda-java-events</artifactId>
             <version>3.4.0</version>
             <scope>provided</scope>
+            <exclusions>
+                <exclusion>
+                    <artifactId>joda-time</artifactId>
+                    <groupId>joda-time</groupId>
+                </exclusion>
+            </exclusions>
```
Comment (ee-usgs) [inline pom.xml:76]: All these exclusions were due to multiple dependencies each including different versions of child libraries as dependencies.  With these changes, there are no longer any version conflicts.

File: pom.xml [hunk 2]
```diff
@@ -93,6 +105,7 @@
         <dependency>
             <groupId>org.postgresql</groupId>
             <artifactId>postgresql</artifactId>
+            <version>42.2.16</version>
```
Comment (ee-usgs) [inline pom.xml:108]: Somehow there was no version spec'ed for this - I didn't realize Maven would let it compile in that case.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWater.java [hunk 3]
```diff
@@ -28,7 +28,7 @@
 	//lev_acy_cd
 	protected String     levelAccuracyCode;
 	//lev_status_cd
-	protected String     siteStatusCode;
+	protected String     readingQualifiers;
```
Comment (ee-usgs) [inline src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWater.java:31]: This field now initially holds the multi-valued JSON string from Aquarius, then when the rules are applied, it is converted to a single digit code that will be used by NWISWeb.  The new digit codes are a new thing - they were single letters before.

File: src/main/resources/sql/selectDiscreteGroundWater.sql [hunk 4]
```diff
@@ -5,7 +5,7 @@ select
     d.measurement_source_code,
     d.measuring_agency_code,
     d.level_accuracy_code,
-    d.site_status_code,
+    d.result_measure_qualifiers::varchar,
```
Comment (ee-usgs) [inline src/main/resources/sql/selectDiscreteGroundWater.sql:8]: I thought about trying to select a single value from the JSON array right in the SQL, but I think it makes sense to use the ordering provided by the enum, which would be really verbose to do in the SQL.  It also makes it easier to see the logic.  
Note:  The actual preference order of qualifier codes is not specified, but I think it at least needs to be repeatable.  PostGres JSONB does not ensure an order to its values.

File: src/test/java/gov/usgs/wma/waterdata/groundwater/JsonType.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+	@Override
+	public void setSqlValue(Object value, int column, PreparedStatement statement) throws SQLException {
+		if (value == null || "".equals(value)) {
+			statement.setNull(column, Types.OTHER);
+		} else {
+			statement.setObject(column, value.toString(), Types.OTHER);
+		}
+	}
+}
```
Comment (ee-usgs) [inline src/test/java/gov/usgs/wma/waterdata/groundwater/JsonType.java:39]: I needed to add this class (copied from the Discrete GW Loader) to be able to process the JSONB type in the tests.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+							if (!StringUtils.isEmpty(val)) {
+								aqQuals.add(StringUtils.trimWhitespace(parser.getValueAsString()));
+							}
+						}
+					}
+
+					if (aqQuals.size() > 0) {
+
+						LevelStatusCode lsc = Stream.of(LevelStatusCode.values())
+													  .filter(a -> a.isMapped())
```
Comment (ssoper-usgs) [inline src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java:60]: Just some whitespace issues here, this first filter seems like it's way over to the right in the git diff.

File: src/test/resources/application.yml [hunk 7]
```diff
@@ -16,10 +16,11 @@ spring:
 logging:
   level:
     root: ${ROOT_LOG_LEVEL}
+    org.postgresql: DEBUG
```
Comment (ssoper-usgs) [inline src/test/resources/application.yml:19]: Just wanted to point this out to make sure you want to keep this set as DEBUG.  Another thing we could do is add this as an environment variable like the line above: `ROOT_LOG_LEVEL` .  That way, we can adjust the postgres logging level from the aws management console on the fly without having to submit PRs.

**Review by ssoper-usgs — APPROVED** [review]

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java (hunk 6 above)
Comment (ee-usgs) [reply src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java:60]: This file is all indented w/ tabs.
...and it looks like most of the other files in this project are tabs as well.  The actual indent levels are correct.

File: src/test/resources/application.yml (hunk 7 above)
Comment (ee-usgs) [reply src/test/resources/application.yml:19]: This is in the /test/.../application.yml, so its not the prod one.  The prod one looks ok.  As a reference, this matches the settings in the same file of the discrete loader.

Comment (ssoper-usgs) [reply src/test/resources/application.yml:19]: Ah, roger that.  I didn't see the `/test` - in that case I think this is a really good default to have for devs running the project locally.

**Comment by duselmann** [comment]
Should sample.rdb be in the project root? Might the test/resources dir be a better path?

**Comment by ee-usgs** [comment]
@duselmann Ha!  No, the sample file wasn't meant to be in there - Nice catch.
I just deleted it.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java (hunk 6 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java:60]: yes, unfortunately our tooling is not consistent with a 4 space tab. It's annoying but until we get over the 8 space tab it will always be an issue.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+					while (! parser.isClosed()) {
+
+						JsonToken token = parser.nextToken();
+
+						if (token != null && token.isScalarValue()) {
+							String val = StringUtils.trimWhitespace(parser.getValueAsString());
+
+							if (!StringUtils.isEmpty(val)) {
+								aqQuals.add(StringUtils.trimWhitespace(parser.getValueAsString()));
```
Comment (duselmann) [inline src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java:52]: I know that tests would be passing but I have a question about token vs parser. It seems odd that we get the value off the parser when we just obtained a current token.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+							}
+						}
+					}
+
+					if (aqQuals.size() > 0) {
+
+						LevelStatusCode lsc = Stream.of(LevelStatusCode.values())
```
Comment (duselmann) [inline src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java]: Please put the stream in the enum as a static method so the code looks something like this
`LevelStatusCode.stream()`

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+													  .filter(a -> a.isMapped())
+								                      .filter(a -> aqQuals.contains(a.getAqDescription()))
+								                      .findFirst().orElse(null);
```
Comment (duselmann) [inline src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java]: Since you are using the stream-lambda pattern, if you returned an Optional<LevelStatusCode> then you could check for isPresent rather than null.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+					}
+
+				} catch (IOException e) {
+					throw new RuntimeException(e);
+				}
+
+			}
+
+			domObj.readingQualifiers = newQualStr;
```
Comment (duselmann) [inline src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java:76]: It is not clear to me why this code goes through enum gyrations to only return a string in the end. I think it should return an enum, no?

File: src/main/resources/sql/selectDiscreteGroundWater.sql (hunk 4 above)
Comment (duselmann) [reply src/main/resources/sql/selectDiscreteGroundWater.sql:8]: AQTS probably does not guaranty an order either so I think we are ok with JSONB lack of order.

File: src/test/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRowHandlerTest.java [hunk 12]
```diff
@@ -66,7 +66,7 @@ DiscreteGroundWater makeDgw() {
 		dgw.measurementSourceCode = "";
 		dgw.measuringAgencyCode = "USGS";
 		dgw.levelAccuracyCode = "2"; // two digits after decimal point
-		dgw.siteStatusCode = ""; // R, S or blank
+		dgw.readingQualifiers = ""; //Numeric code
```
Comment (duselmann) [inline src/test/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRowHandlerTest.java:69]: Wait, isn't this a JSON object, not a numeric?

Comment (duselmann) [reply src/test/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRowHandlerTest.java:69]: Oh, now I get it -- because of unit tests.

**Review by duselmann — APPROVED** [review]
I have some suggestions but overall is good to go.

File: src/test/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRowHandlerTest.java (hunk 12 above)
Comment (ssoper-usgs) [reply src/test/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRowHandlerTest.java:69]: @duselmann I was confused initially too, but I think NWISWeb expects a numeric code here, so it's just mapped based on those incoming quals.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java (hunk 8 above)
Comment (ee-usgs) [reply src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java:52]: The syntax of the Jackson parser is confusing.  JsonToken is just an enum, intended to tell you what what the next 'thing' is.  nextToken() advances the pointer and gives you the token, but to actually get the content at the pointer, 'get' needs to be called on the parser.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java (hunk 10 above)
Comment (ee-usgs) [reply src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java]: OK, I think that is slightly clearer - I'll add that.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java (hunk 11 above)
Comment (ee-usgs) [reply src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java:76]: Yeah, good question, and one I think is worth discussing b/c it gets a the basic pattern of how we ETL values.
What we get from AQTS is a JSON String containing an array of values.  What we need to put into the RDB is a single digit.  The incoming JSON String might be a bit negotiable, but that is currently what is preserved all the way from AQTS into the Observations DB.

One simple way to think of this is that the enum is a stand-in for a DAO.  For now the list is considered unchanging, but the enum could be thought of as just a structured hack to hardcode what might have been a db look-up.

I want to go a little further, though, because there are cases where we use enums for simple, unchanging value lists.  First, my key goal in these ETLs is to separate and consolidate the Business Logic for several reasons:
* By separating BL from other code, I can more easily unit test it without having to do IT tests w/ db CSV files to create edge cases.  Unit tests let me really pound the BL w/ every weird edge case I can think of.
* By consolidating the BL, it's much easier to review what rules (validations, mappings, calculations) are in place.

With that in mind, I have a Rules object that processes a Domain object applying the BL rules.  The dom object needs to be able to hold the AQTS value (String) and a representation of the output value.  If I want to put an enum into the dom obj, I could:
* Add a separate field to hold the enum or
* I could have two dom objs (separate AQTS and NWISWeb ones)
I think both options are more complex than needed.  My source value and destination value are both Strings, so it works to just modify the existing value according to the rules.

I also think that it is slightly preferable to not put an enum in the dom obj when following this 'rules' pattern, since it means that this field becomes special, requiring some special code to dereference the enum to its string representation.  That results in just a little bit of logic when we write from the dom obj.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java (hunk 9 above)
Comment (ee-usgs) [reply src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java]: I'm not sure why this is better.  I would just move this same code to a new method on the enum resulting in more code.

**Comment by ee-usgs** [comment]
Given that both of you approve of these changes and I want to base my IOW-666 code on this code, I'm going to merge it now.  There are some ongoing discussions, but I can incorporate any further changes into my work on the '666' ticket.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java (hunk 10 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java]: Optionals are the recommended pattern for streams that might not return a result. Thanks.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java (hunk 9 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java]: it is cleaner to me to have it as I suggested.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java (hunk 11 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java:76]: I think you explained the part I understand and not the part I am unclear about. We discussed the enum mechanism together  as a DOA substitute. What I am unclear about is that an enum is not returned once it is determined. Why a string?

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java (hunk 9 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java]: when we use the streams features we call collection.stream()..... it would be nice if that pattern was true for this enum too.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java (hunk 8 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/waterdata/groundwater/DiscreteGroundWaterRules.java:52]: oh, that is a strange concept. no, actually I think it is a misnamed method. not that you should change the lib, but that method should be something like currentTokenType()
